src/AI/JakeMan.java

```diff
@@ -0,0 +1,770 @@
+    public Travel(Army co, JakeMan ai)
+    {
+      super(co, ai);
+      this.ai = ai;
+    }
+
+    @Override
+    public GameAction getUnitAction(Unit unit, GameMap gameMap)
+    {
+      ai.log(String.format("Evaluating travel for %s.", unit.toStringWithLocation()));
+      boolean avoidProduction = false;
+      return ai.findTravelAction(gameMap, unit, false, avoidProduction);
+    }
+  }
+
+  public static class BuildStuff implements AIModule
+  {
+    private static final long serialVersionUID = 1L;
+    public final Army myArmy;
+    public final JakeMan ai;
+
+    public BuildStuff(Army co, JakeMan ai)
+    {
+      myArmy = co;
+      this.ai = ai;
+    }
+
+    Map<XYCoord, UnitModel> builds;
+
+    @Override
+    public void endTurn()
+    {
+      if( null != builds )
+      {
+        ai.log(String.format("Warning - builds not null on turn end; contains %s", builds));
+        builds = null;
+      }
+    }
+
+    @Override
+    public GameAction getNextAction(PriorityQueue<Unit> unitQueue, GameMap gameMap)
+    {
+      if( null == builds )
+        builds = ai.queueUnitProductionActions(gameMap);
+
+      for( XYCoord coord : new ArrayList<XYCoord>(builds.keySet()) )
+      {
+        ai.log(String.format("Attempting to build %s at %s", builds.get(coord), coord));
+        Unit resident = gameMap.getResident(coord);
+        if( null != resident )
+        {
+          boolean avoidProduction = true;
+          GameAction eviction = null;
+          if( resident.CO.army == myArmy && !resident.isTurnOver )
+            eviction = ai.evictUnit(gameMap, null, resident, avoidProduction);
+          if( null != eviction )
+            return eviction;
+          else
+          {
+            ai.log(String.format("  Can't evict unit %s to build %s", resident.toStringWithLocation(), builds.get(coord)));
+            builds.remove(coord);
+            continue;
+          }
+        }
+        MapLocation loc = gameMap.getLocation(coord);
+        Commander buyer = loc.getOwner();
+        ArrayList<UnitModel> list = buyer.getShoppingList(loc);
+        UnitModel toBuy = builds.get(coord);
+        if( buyer.getBuyCost(toBuy, coord) <= myArmy.money && list.contains(toBuy) )
+        {
+          builds.remove(coord);
+          return new GameAction.UnitProductionAction(buyer, toBuy, coord);
+        }
+        else
+        {
+          ai.log(String.format("  Trying to build %s, but it's unavailable at %s", toBuy, coord));
+          continue;
+        }
+      }
+
+      builds = null;
+      return null;
+    }
+  }
+
+  /** Produces a list of destinations for the unit, ordered by their relative precedence */
+  private ArrayList<XYCoord> findTravelDestinations(
+                                  GameMap gameMap,
+                                  Unit unit,
+                                  boolean avoidProduction )
+  {
+    UnitContext uc = new UnitContext(gameMap, unit);
+    uc.calculatePossibleActions();
```

Comment:
Is this just to populate the field inside the `UnitContext`? Intuitive.

How is that different/better than `AIUtils.getAvailableUnitActions()`? I feel like, at some level, we are just implementing parallel APIs.

Reply:
`calculatePossibleActions()` populates the set of action *types* your unit can do based on any `UnitModifier`s it has.
Is there a better name for it?

Reply:
Oh, gotcha. Maybe `calculatePossibleActionTypes()`. I might even go crazy and rename `UnitContext.possibleActions` to `UnitContext.actionFactories/Types` and change this function to `populateActionFactories/Types`.